File: appactions/interaction/interaction-capabilities-core/src/main/java/androidx/appactions/interaction/capabilities/core/values/properties/Participant.java
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.appactions.interaction.capabilities.core.values.properties;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appactions.interaction.capabilities.core.values.Person;

import com.google.auto.value.AutoValue;

import java.util.Optional;

/**
 * Represents the value of the union property: http://schema.org/participant, currently it only can
 * contain {@link Person}.
 */
public class Participant {
    private final Value mValue;

    public Participant(@NonNull Person person) {
        mValue = new AutoValue_Participant_Value(Optional.of(person));
    }

    @NonNull
    public Optional<Person> asPerson() {
        return mValue.person();
    }

    @Override
    public int hashCode() {
        return mValue.hashCode();
    }

    @Override
    public boolean equals(@Nullable Object object) {
        if (object instanceof Participant) {
            Participant that = (Participant) object;
            return this.mValue.equals(that.mValue);
        }
        return false;
    }

    /** Represents the value in the this wrapper class. */
    @AutoValue
    abstract static class Value {
        abstract Optional<Person> person();
    }
}

File: core/core-telecom/src/main/java/androidx/core/telecom/extensions/Participant.kt
/*
 * Copyright 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.core.telecom.extensions

import androidx.core.telecom.util.ExperimentalAppActions

@ExperimentalAppActions
internal fun ParticipantParcelable.toParticipant(): Participant {
    return Participant(id, name)
}

/**
 * A representation of aspects of a participant in a Call.
 *
 * @param id A unique identifier shared with remote surfaces that represents this participant. This
 *   value MUST be unique and stable for the life of the call, meaning that this ID should not
 *   change or be reused for the lifetime of the call.
 * @param name The name of the Participant, which remote surfaces will display to users.
 */
@ExperimentalAppActions
public class Participant(
    public val id: String,
    public val name: CharSequence,
) {

    internal fun toParticipantParcelable(): ParticipantParcelable {
        return ParticipantParcelable().also { parcelable ->
            parcelable.id = id
            parcelable.name = name
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Participant

        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun toString(): String {
        return "Participant[$id]: name=$name"
    }
}
